hbase-compression/hbase-compression-aircompressor/src/main/java/org/apache/hadoop/hbase/io/compress/aircompressor/SnappyCodec.java

```diff
@@ -0,0 +1,144 @@
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import io.airlift.compress.snappy.SnappyCompressor;
+import io.airlift.compress.snappy.SnappyDecompressor;
+
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeys;
+import org.apache.hadoop.io.compress.BlockCompressorStream;
+import org.apache.hadoop.io.compress.BlockDecompressorStream;
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.apache.hadoop.io.compress.CompressionInputStream;
+import org.apache.hadoop.io.compress.CompressionOutputStream;
+import org.apache.hadoop.io.compress.Compressor;
+import org.apache.hadoop.io.compress.Decompressor;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Hadoop snappy codec implemented with aircompressor.
+ * <p>
+ * This is data format compatible with Hadoop's native snappy codec.
+ */
+@InterfaceAudience.Private
+public class SnappyCodec implements Configurable, CompressionCodec {
+
+  private Configuration conf;
+
+  public SnappyCodec() {
+    conf = new Configuration();
+  }
+
+  @Override
+  public Configuration getConf() {
+    return conf;
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public Compressor createCompressor() {
+    return new HadoopSnappyCompressor();
+  }
+
+  @Override
+  public Decompressor createDecompressor() {
+    return new HadoopSnappyDecompressor();
+  }
+
+  @Override
+  public CompressionInputStream createInputStream(InputStream in) throws IOException {
+    return createInputStream(in, createDecompressor());
+  }
+
+  @Override
+  public CompressionInputStream createInputStream(InputStream in, Decompressor d)
+      throws IOException {
+    return new BlockDecompressorStream(in, d, getBufferSize(conf));
+  }
+
+  @Override
+  public CompressionOutputStream createOutputStream(OutputStream out) throws IOException {
+    return createOutputStream(out, createCompressor());
+  }
+
+  @Override
+  public CompressionOutputStream createOutputStream(OutputStream out, Compressor c)
+      throws IOException {
+    int bufferSize = getBufferSize(conf);
+    int compressionOverhead = (bufferSize / 6) + 32;
+    return new BlockCompressorStream(out, c, bufferSize, compressionOverhead);
+  }
+
+  @Override
+  public Class<? extends Compressor> getCompressorType() {
+    return HadoopSnappyCompressor.class;
+  }
+
+  @Override
+  public Class<? extends Decompressor> getDecompressorType() {
+    return HadoopSnappyDecompressor.class;
+  }
+
+  @Override
+  public String getDefaultExtension() {
+    return ".snappy";
+  }
+
+  private static int getBufferSize(Configuration conf) {
+    int size = conf.getInt(
+      CommonConfigurationKeys.IO_COMPRESSION_CODEC_SNAPPY_BUFFERSIZE_KEY,
+      CommonConfigurationKeys.IO_COMPRESSION_CODEC_SNAPPY_BUFFERSIZE_DEFAULT);
+    return size > 0 ? size : 256 * 1024;
+  }
+
+  @InterfaceAudience.Private
```

Comment:
My understanding is we do it, but I'm also not sure of the best practice.

Reply:
I can just remove these, I actually don't remember what I was doing here. Was on the first day of work, something like that...

Reply:
IIRC we started expressly labeling them because there was some uncertainty in the community on wether or not nested classes were implicitly covered by their containing class's audience.

Reply:
Now that I think about it more, I wonder if it's a limitation of the audience annotation tooling's ability to check annotations on containing classes.

Reply:
@busbey @virajjasani I will label nested classes as private to be sure.

Reply:
Sounds good.